Docs/auth flows and detail (#1161)
canonical/testflinger: docs/explanation/authentication.rst

.. _authentication:

> Authentication and Authorisation

Author: ```suggestion
Authentication and Authorization
```

Reviewer: Its been a while since I read the recommendations, but as far as I remember, UK spelling was preferred over US spelling. @tang-mm perhaps you may know which is the latest recommendation? I honestly don't remember where the documentation was

Author: There was a recent (since I've been here) announcement that US English was to be used for new projects.  There aren't many of these, and we can leave this one until and if we decide to do the whole thing; but I do like to see my `z`'s when I can.

Reviewer: Yeah, once we clarify this we can check which other words are in UK English and replace them... I saw the UK recommendation when I was starting at Canonical so I might be the one outdated!

Author: no change for now.

The following is the simplified priority order of authentication methods:

1. Basic Auth using ``client_id`` and ``secret_key``
2. Bearer token using a refresh token
3. OIDC flow if enabled on the server

Authentication Flows
~~~~~~~~~~~~~~~~~~~~

The following sequence diagrams show the authentication flow for each of the
authentication methods and the interactions between the user, the CLI and the
server.

Basic Auth
^^^^^^^^^^

Regardless of whether OIDC is enabled or not, Basic Auth will be used if both the
``client_id`` and ``secret_key`` are provided.

.. mermaid::
    :alt: Sequence diagram showing BASIC Auth is used when client_id and secret_key are provided
    :caption: Sequence diagram showing BASIC Auth is used when client_id and secret_key are provided
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Have client-id + secret-key, use basic auth
        User->>CLI: Run command with client-id + secret-key
        CLI->>CLI: Validate both parameters are present

>         CLI->>Server: Request with Basic Auth<br/>base64(client-id:secret-key)

Author: I would like to add the endpoint, since it is explicitly the authorization endpoint: /oauth2/token

        Server->>Server: Validate client credentials
        Server-->>CLI: 200 OK / return access and refresh tokens

>         CLI-->>Server: use access token to complete the request

Author: In this case, we would send the request to the intended endpoint, e.g. `/jobs/<job_id>`; the "Validate client permissions" could be:
a) Validate token authenticity (token sig)
b) Validate permissions for endpoint

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli

        Note over User,CLI: One of client-id or secret-key is missing
        User->>CLI: Run command with wrong number of credential arguments
        CLI-->>User: Show usage

Refresh Token
^^^^^^^^^^^^^

If client credentials are not provided, the CLI will attempt to use a refresh
token if one is available. The refresh token is stored in a snap-local cache and is
used to obtain a new access token for the user. The refresh token is valid for 6 days,
after which it will expire and the user will need to re-authenticate.

When a valid refresh token is present, this flow applies regardless of whether
OIDC is enabled on the server:

.. mermaid::
    :alt: Sequence diagram showing a refresh token being used and accepted
    :caption: Sequence diagram showing a refresh token being used and accepted
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Neither client-id nor secret-key, refresh token accepted
        User->>CLI: Run command without client-id + secret-key

>         CLI->>CLI: Load stored refresh/bearer token

Author: ```suggestion
        CLI->>CLI: Load stored refresh token
```

>         CLI->>Server: Request with Bearer refresh token

Author: ```suggestion
        CLI->>Server: Request with refresh token
```

Author: add a note about which endpoint: `/oauth2/refresh` (vs `token` in the Basic case)

        Server->>Server: Validate token
        Server-->>CLI: 200 OK / return access and refresh tokens

>         CLI-->>Server: use access token to complete the request

Author: add endpoint note; e.g. /jobs/<job_id>

.. mermaid::
    :alt: Sequence diagram showing a refresh token being rejected and the server initiating OIDC process
    :caption: Sequence diagram showing a refresh token being rejected and the server initiating OIDC process
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server
        participant OIDC as OIDC provider

        Note over User,OIDC: Neither client-id nor secret-key, token rejected, server uses OIDC
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh/bearer token
        CLI->>Server: Request with Bearer refresh token
        Server->>Server: Token expired or rejected
        Server-->>OIDC: Initiate OIDC auth flow
        OIDC-->>Server: Auth code
        Server->>CLI: 401 Unauthorized, try this instead: user code + URL for OIDC handshake
        CLI->>CLI: Delete stored token
        CLI->>User: Display URL and code
        User->>OIDC: Complete the handshake
        CLI->>Server: Poll for auth completion
        Server-->>OIDC: Is this user cool?
        OIDC-->>Server: Yes + user identity (email)

>         Server->>CLI: Issue refresh token

Author: ```suggestion
        Server->>CLI: Issue refresh and access token
```

>         CLI->>Server: Retry request with new Bearer token

Author: ```suggestion
        CLI->>Server: Send request with access token
```
Add note about uri -- now we finally use the access token to reach the desired endpoint

    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server
        participant OIDC as OIDC provider

        Note over User,OIDC: Neither client-id nor secret-key, token rejected, server uses OIDC
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh token
        CLI->>Server: Request with refresh token
        Note over CLI,Server: POST /v1/oauth2/refresh
        Server->>Server: Token expired or rejected
        Server-->>CLI: 400 Bad Request
        CLI->>CLI: Delete stored refresh token
        CLI->>Server: Initiate OIDC auth flow
        Note over CLI,Server: POST /oidc/auth-init
        Server-->>OIDC: Proxy request to OIDC provider
        OIDC-->>Server: device_code + user_code + URL for OIDC handshake
        Server->>CLI: request_id + user code + URL for OIDC handshake
        CLI->>User: Display URL and code
        User->>OIDC: Complete the handshake
        CLI->>Server: Poll for auth completion
        Note over CLI,Server: POST /oidc/auth-poll/<request_id>
        Server-->>OIDC: Is this user cool?

>         OIDC-->>Server: Yes + user identity (email)

Reviewer 2: ```suggestion
        Server-->>OIDC: Validate authorization status
        OIDC-->>Server: Authorized + user identity (email)
```
Let's update this wording to use a more formal tone

''''''''''''''''''''''''''''''

.. mermaid::
    :alt: Sequence diagram showing no stored token with OIDC enabled, triggering the OIDC flow
    :caption: Sequence diagram showing no stored token with OIDC enabled, triggering the OIDC flow
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server
        participant OIDC as OIDC provider

        Note over User,OIDC: Neither client-id nor secret-key, no token, server uses OIDC
        User->>CLI: Run command without credentials or token
        CLI->>CLI: No stored bearer/refresh token found
        CLI->>Server: Initiate OIDC auth flow
        Note over CLI,Server: POST /oidc/auth-init
        Server-->>OIDC: Proxy request to OIDC provider
        OIDC-->>Server: device_code + user_code + URL for OIDC handshake
        Server->>CLI: request_id + user code + URL for OIDC handshake
        CLI->>User: Display URL and code
        User->>OIDC: Complete the handshake
        CLI->>Server: Poll for auth completion
        Note over CLI,Server: POST /oidc/auth-poll/<request_id>
        Server-->>OIDC: Is this user cool?

>         OIDC-->>Server: Yes + user identity (email)

Reviewer 2: ```suggestion
        Server-->>OIDC: Validate authorization status
        OIDC-->>Server: Authorized + user identity (email)
```
same

.. _authentication:

Authentication and Authorisation
================================

Authentication
--------------

Testflinger can be configured to use OpenID Connect (OIDC) for authentication. The way that
authorization working in the Testflinger server depends on whether OIDC is enabled or not. When
OIDC is *not* enabled, Testflinger supports anonymous logins. However, when OIDC is enabled
each request must be authenticated. Both operating modes support use of self-signed tokens

> for authentication, and these are what are used in typical CI/CD flows.

Reviewer: I feel that this is not accurate for CI/CD flows as "self-signed tokens" is what happens under the hood. Also, I feel like we perhaps should not need to specify that those are for CI/CD as this is mostly related on how we use it. 

Perhaps more accurate would be something like this:

> Both operating modes supports the use of a Testflinger ``client_id`` and ``secret_key`` pair that can be generated by an administrator and used as part of the Basic Auth. 

This also relates directly to the next paragraph, what do you think?

Author: ```suggestion
each request must be authenticated.  Both operating modes support use of basic auth accounts for authentication which are suitable for use with CI/CD flows.
```

Reviewer: The rewording makes sense but still the "self-signed" tokens is the one its bugging me a bit. Specially because the only signed is the `access_token` and CI/CD flows don't use this

Since Basic Auth can be used in both configurations, it is prioritized first, when

> Basic Auth is used it will be handled within the server before other authentication methods.

Reviewer: I think we can remove one mentioned to the Basic Auth, something like:

```suggestion
Since Basic Auth can be used in both configurations, it is prioritized and will be handled within the server before other authentication methods.
```

Authentication Flows
--------------------

The following sequence diagrams show the authentication flow for each of the
authentication methods and the interactions between the user, the CLI and the
server.

.. note::

    For simplicity, all of the diagrams are targetting the ``/v1/jobs`` endpoint
    on Testflinger server side but the same flow applies to any other endpoint
    that requires authentication.

The Testflinger server can be configured one of two ways:
 - Without OIDC Authentication Configured (OIDC Disabled)
 - With OIDC Authentication Configured (OIDC Enabled)

OIDC Disabled
~~~~~~~~~~~~~

When Testflinger is configured to *not* use OIDC for authentication, the order is:

1. Basic Auth using ``client_id`` and ``secret_key``
2. Bearer token using a refresh token
3. Any non-authenticated requests are treated as an anonymous CONTRIBUTOR role.

.. _basic-auth-is-oidc-agnostic:

Basic Auth
''''''''''

Regardless of whether OIDC is enabled or not, Basic Auth will be used if both the
``client_id`` and ``secret_key`` are provided.

.. mermaid::

>     :alt: Sequence diagram showing BASIC Auth is used when client_id and secret_key are provided

Author: ```suggestion
    :alt: Sequence diagram showing Basic Auth is used when client_id and secret_key are provided
```

>     :caption: Sequence diagram showing BASIC Auth is used when client_id and secret_key are provided

Author: ```suggestion
    :caption: Sequence diagram showing Basic Auth is used when client_id and secret_key are provided
```

OIDC is enabled on the server:

.. mermaid::
    :alt: Sequence diagram showing a refresh token being used and accepted
    :caption: Sequence diagram showing a refresh token being used and accepted
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Neither client-id nor secret-key, refresh token accepted
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh token
        CLI->>Server: Request with refresh token
        Note over CLI,Server: POST /v1/oauth2/refresh
        Server->>Server: Validate token
        Server-->>CLI: 200 OK / return access and refresh tokens
        CLI-->>Server: use access token to complete the request
        Note over CLI,Server: POST /v1/jobs
        Server->>Server: Validate access token signature and client permissions
        Server-->>CLI: 200 OK / authenticated response
        CLI-->>User: Show result

Token expired or rejected
.........................

If the server does not have OIDC enabled and the refresh token is rejected,

> the server returns a 401 and the CLI removes the stored token. If the CLI does not

Reviewer: Perhaps it doesn't hurt, but I think this is now redundant given we are the "OIDC disabled" section. 

It can perhaps directly be simplified to:

```suggestion
The server returns a 401 and the CLI removes the stored token. If the CLI does not
```

This is different as the cases above where I think it has some value to say that it works regardless OIDC is enabled or not as we are referencing in the below.

automatically retry the operation anonymously it will signal to the user that the
request was denied so the user may try again.

.. mermaid::
    :alt: Sequence diagram showing a refresh token being rejected and the user being denied access
    :caption: Sequence diagram showing a refresh token being rejected and the user being denied access
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Neither client-id nor secret-key, token rejected, server does not use OIDC
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh token
        CLI->>Server: Request with refresh token
        Note over CLI,Server: POST /v1/oauth2/refresh
        Server->>Server: Token expired or rejected
        Server-->>CLI: 400 Bad Request
        CLI->>CLI: Delete stored refresh token
        CLI-->>User: Authentication failed

No credentials or stored token
''''''''''''''''''''''''''''''

If the server does not require authentication, the request is allowed through

> as an anonymous user:

Reviewer: Given this is a OIDC disabled section, this can also be reworded/simplified but I don't think as necessarily as the other [comment](https://github.com/canonical/testflinger/pull/1161/changes#r3617801912)

Author: I prefer to keep the verbosity as these are long sections and being able to glance down and have the paragraphs clearly stating with OIDC, etc.  In this case, I agree that it can be re-worded, but I would only make it more technically correct:

```suggestion
If a given endpoint does not require authentication, the request is allowed through
as an anonymous user:
```

Reviewer: Alright, I can be sold that idea and the rewording makes sense and its more accurate, if you are to "Apply suggestion" just note that there is a "the" before the "a given endpoints", perhaps this is

> If a given endpoint

.. mermaid::
    :alt: Sequence diagram showing no auth for anonymous login
    :caption: Sequence diagram showing no auth for anonymous login
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Neither client-id nor secret-key, no token, server does not require auth
        User->>CLI: Run command without credentials or token
        CLI->>CLI: No stored bearer/refresh token found
        CLI->>Server: Request without Authorization header
        Note over CLI,Server: POST /v1/jobs
        Server->>Server: Auth not required
        Server-->>CLI: 200 OK / anonymous user response
        CLI-->>User: Show result

OIDC Enabled
~~~~~~~~~~~~

When Testflinger is configured to use OpenID Connect (OIDC) for authentication, the order is:

> When Testflinger is configured to use OpenID Connect (OIDC) for authentication, the order is:

Reviewer: ```suggestion
```

1. Basic Auth using ``client_id`` and ``secret_key``
2. Bearer token using a refresh token
3. OIDC authentication flow

Basic Auth
''''''''''

Exactly the same as without OIDC, see :ref:`basic-auth-is-oidc-agnostic`.

Refresh Token
'''''''''''''

Exactly the same as without OIDC, see :ref:`refresh-auth-is-oidc-agnostic`.

Token expired or rejected
.........................

If the server has OIDC enabled and the refresh token is rejected, the server

> initiates an OIDC authentication flow:

Reviewer: Same as this [comment](https://github.com/canonical/testflinger/pull/1161/changes#r3617801912) perhaps redundant to say it has OIDC enabled  as we are already in this section

Author: I like the verbosity, arguments above.

Author: If you want to propose dropping the second "OIDC" I can do that.  Please add your comment/suggestion

Reviewer: Its okay, if you prefer to keep the initial OIDC the second also makes sense. I can only see `the authentication flow` but its more explicit as-is to indicate this starts the OIDC-specific process

.. tip::
    You can also run ``testflinger-cli login`` without command line arguments if your credentials
    are located in a ``.env`` file as mentioned in :doc:`Authentication using Testflinger CLI <../how-to/authentication>`

Authorization Flows Based on Configuration and Provided Auth Parameters

> =======================================================================

Reviewer 2: Is this heading supposed to be an H2 heading beneath the same title? The current marker is at the same level as "Authentication and Authorisation", which makes the document appear as a new topic in the sidebar, and thus the sub-headings are not rendered in the right side panel in the [preview](https://canonical-ubuntu-documentation-library--1161.com.readthedocs.build/testflinger/1161/explanation/authentication/#)

I'd assume this level should use dash `--------`, and H3-heading with `~~~`: 

```
Authorization Flows Based on Configuration and Provided Auth Parameters
------------------------------------------------------------------------

When the client_id and secret_key are provided:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reviewer: I modified all the headings so this should be already fixed

        CLI->>Server: Request with Basic Auth<br/>base64(client-id:secret-key)
        Server->>Server: Validate client credentials
        Server-->>CLI: 200 OK / authorized response
        CLI-->>User: Show result

When the client_id and secret_key are NOT provided:
---------------------------------------------------

When a refresh token is available:
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. mermaid::
    :alt: Sequence diagram showing a refresh token being used and accepted
    :caption: Sequence diagram showing a refresh token being used and accepted
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Neither client-id nor secret-key, refresh token accepted
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh/bearer token
        CLI->>Server: Request with Bearer refresh token
        Server->>Server: Validate token
        Server-->>CLI: 200 OK / authenticated as token-associated user
        CLI-->>User: Show result
    
When OIDC is enabled:
^^^^^^^^^^^^^^^^^^^^^

> When the refresh token is not accepted:

Reviewer 2: reword to something like:
```suggestion
Token refresh fails
```

Reviewer: I think this is now: `Token expired or rejected`. I'll mark this as resolved for now, but please let me know if that also needs a clarification

.......................................

.. mermaid::
    :alt: Sequence diagram showing a refresh token being rejected and the server initiating OIDC process
    :caption: Sequence diagram showing a refresh token being rejected and the server initiating OIDC process
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

> 

Author: @copilot; update from OIDC to Server, don't add participants.

        Note over User,OIDC: Neither client-id nor secret-key, token rejected, server uses OIDC
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh/bearer token
        CLI->>Server: Request with Bearer refresh token
        Server->>Server: Token expired or rejected
        Server-->>OIDC: Initiate OIDC auth flow
        OIDC-->>Server: Auth code
        Server->>CLI: 401 Unauthorized, try this instead: Auth code
        CLI->>CLI: Delete stored token
        CLI->>User: Display URL and code
        User->>OIDC: Complete the handshake
        CLI->>Server: Poll for auth completion
        Server-->>OIDC: Is this user cool?
        OIDC-->>Server: Yes + user identity (email)
        Server->>CLI: Issue refresh token
        User-->>CLI: Potential user re-issue of command?
        CLI->>Server: Retry request with new Bearer token
        Server-->>CLI: 200 OK / authenticated response
        CLI-->>User: Show result
   

> When the refresh token is not available:

Reviewer 2: ```suggestion
First-time login or missing token
```

Reviewer: I think this is now: `No credentials or stored token` I'll mark this as resolved for now, but please let me know if this perhaps is clearer or preferred as suggested!

........................................

.. mermaid::
    :alt: Sequence diagram showing a refresh token being used and accepted
    :caption: Sequence diagram showing a refresh token being used and accepted

>     :align: center

Author: @copilot: revert all your changes to all files.

        Server->>CLI: Issue refresh token
        User-->>CLI: Potential user re-issue of command?
        CLI->>Server: Retry request with new Bearer token
        Server-->>CLI: 200 OK / authenticated response
        CLI-->>User: Show result

When OIDC is NOT enabled:
^^^^^^^^^^^^^^^^^^^^^^^^^

When the refresh token is not accepted:
.......................................

.. mermaid::
    :alt: Sequence diagram showing a refresh token being rejected and the user being denied access
    :caption: Sequence diagram showing a refresh token being rejected and the user being denied access
    :align: center

    sequenceDiagram
        actor User
        participant CLI as testflinger-cli
        participant Server as testflinger-server

        Note over User,Server: Neither client-id nor secret-key, token rejected, server does not use OIDC
        User->>CLI: Run command without client-id + secret-key
        CLI->>CLI: Load stored refresh/bearer token
        CLI->>Server: Request with Bearer refresh token
        Server->>Server: Token expired or rejected
        Server-->>CLI: 401 Unauthorized, delete stored token
        CLI->>CLI: Delete stored token

>         CLI-->>User: Authentication faile

Author: ```suggestion
        CLI-->>User: Authentication failed
```
